samza-core/src/main/java/org/apache/samza/coordinator/StreamRegexMonitor.java

```diff
@@ -0,0 +1,232 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.coordinator;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Sets;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import org.apache.samza.metrics.Gauge;
+import org.apache.samza.metrics.MetricsRegistry;
+import org.apache.samza.system.StreamMetadataCache;
+import org.apache.samza.system.SystemStream;
+import org.apache.samza.system.SystemStreamMetadata;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scala.collection.JavaConverters;
+
+
+/**
+ * Periodically monitors input regexes
+ */
+public class StreamRegexMonitor {
+  private static final Logger log = LoggerFactory.getLogger(StreamRegexMonitor.class);
+
+  private enum State { INIT, RUNNING, STOPPED }
+
+  private final Set<SystemStream> streamsToMonitor;
+  private final Map<String, Pattern> systemRegexesToMonitor;
+  private final StreamMetadataCache metadataCache;
+  private final int inputRegexMonitorPeriodMs;
+
+  // Map of gauges (one per system), emitted when new input stream for that system is detected
+  private final Map<String, Gauge<Integer>> gauges;
+
+  private final Callback callbackMethod;
+
+  // Used to guard write access to state.
+  private final Object lock = new Object();
+
+  // Factory of daemon-threads to create the single threaded executor pool
+  private static final ThreadFactory THREAD_FACTORY = new ThreadFactoryBuilder().setDaemon(true)
+      .setNameFormat("Samza-" + StreamRegexMonitor.class.getSimpleName())
+      .build();
+  private final ScheduledExecutorService schedulerService = Executors.newSingleThreadScheduledExecutor(THREAD_FACTORY);
```

Comment:
group static and non-static fields together. 

personally prefer statics first, then instance fields